Sources/ClientRuntime/Networking/Http/Middlewares/FlexibleChecksumsResponseMiddleware.swift

```diff
@@ -0,0 +1,108 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0.
+
+public struct FlexibleChecksumsResponseMiddleware<OperationStackOutput>: Middleware {
+    
+    public let id: String = "FlexibleChecksumsResponseMiddleware"
+    
+    // The priority to validate response checksums, if multiple are present
+    let CHECKSUM_HEADER_VALIDATION_PRIORITY_LIST: [String] = [
+        HashFunction.crc32c,
+        .crc32,
+        .sha1,
+        .sha256
+    ].map { $0.toString() }
+
+    let validationMode: Bool
+
+    public init(validationMode: Bool) {
+        self.validationMode = validationMode
+    }
+
+    public func handle<H>(context: Context,
+                          input: SdkHttpRequest,
+                          next: H) async throws -> OperationOutput<OperationStackOutput>
+    where H: Handler,
+    Self.MInput == H.Input,
+    Self.MOutput == H.Output,
+    Self.Context == H.Context {
+
+        // The name of the checksum header which was validated. If `null`, validation was not performed.
+        context.attributes.set(key: AttributeKey<String>(name: "ChecksumHeaderValidated"), value: nil)
+
+        // Initialize logger
+        guard let logger = context.getLogger() else {
+            throw ClientError.unknownError("No logger found!")
+        }
+        
+        // Exit if validation should not be performed
+        if (!validationMode) {
+            logger.info("Checksum validation should not be performed! Skipping workflow...")
+            return try await next.handle(context: context, input: input)
+        }
+
+        // Get the response
+        let response = try await next.handle(context: context, input: input)
```

Comment:
Wouldn't the response already be received & deserialized by the time we reach this point of this middleware, since we're placing the middleware _after_ the DeserializeMiddleware?

Reply:
Note: Ongoing offline discussion on whether calling `next.handle` here and one more time later means executing request twice & if not, how calling `next.handle` here conveniently gives a copy of the response both times

Reply:
After doing some research, and perhaps someone else can confirm, I do not believe calling next.handle() issues the request multiple times.

Evidence for this:
- The deserializeMiddleware is actually being added to the stack `before` subsequent middlewares such as FlexibleChecksumsResponseMiddleware. The line `let response = try await next.handle(...)` is a recursive call saying "come back to me with the response we get after the entire middleware stack is done". The next call in the stack that happens is going to FlexibleChecksumsResponseMiddleware where we again say `let response = try await next.handle(...)` and ask that the result of calling subsequent middlewares is given back to me here. Then the response should be issued and goes back to FlexibleChecksumsResponseMiddleware. And once we're done there then we go back to DeserializeMiddleware where eventually we return either a `throw try await httpResponseErrorClosure(copiedResponse.httpResponse)` or `return copiedResponse`. While this is **not** the last middleware added to the stack, it ends up being the last middleware since it is the first in the stack to ask for the response back **and** does not end up returning a `try await next.handle(...)` at the end.
- Inspecting the response in both places this is called I can see that the memory address of the stream in httpResponse is same
- Looking at the logs we would expect to see a completion handler execute more than once with the bytes streamed down before it if the same call to pull down a file was executed more than once, but this is not the case

**My Conclusion:** It is ok from a networking perspective to have response held in more than one middleware as long as it is after the DeserializeMiddleware in the stack. However, we are holding this thing in memory so it should be a stream with a known content length (not event streams).